Volume type standard (#351)
SovereignCloudStack/standards: Standards/scs-XXXX-v1-volume-type-standard.md

---
title: Volume Type Standard
type: Standard
status: Draft
track:  IaaS
---

## Introduction

> Volume Types are used to classify volumes and provide a basic decision for what kind of volume should be created. These volume types can sometimes be very backend-specific and it might be hard for a user to choose the most suitable volume type, if there is more than one type.

Reviewer: Comparing this MR with https://github.com/SovereignCloudStack/standards/pull/453/ I prefer MR #453, as it's more precise.

Author: I changed it.

### Glossary

The following special terms are used throughout this standard document:
| Term | Meaning |
|---|---|
| volume | OpenStack ressource, virtual drive which usually resides in a network storage backend |
| AZ | Availability Zone |
| Volume QoS | Quality of Service object for Volumes |

> ## Motivation

Reviewer: I prefer motivation from #453

Author: I changed it.

## Design Considerations

All Considerations can be looked up in detail in the [Decision Record for the Volume Type Standard.](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)

### Systematic Description of Volume Types

To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. As for the time this standard is created, there is no way for users to discover all aspects through OpenStack commands. Therefore the aspects, that are fulfilled within a volume type, should be stated in the beginning of the **description** of a volume type in the following manner:

> `[scs:aspect1][scs:aspect2]...`

Reviewer: IMO, we can omit repetition of prefix `scs`
```suggestion
`[scs:aspect1, aspect2,... aspect2]`
```

Author: i changed it.

There is no sorting of aspects required. Every aspect should only be mentioned to the maximal amount of one.

### Standardized Aspects

The following table shows, which aspects are considered in this standard. The last column shows how the description of the volume type has to be adjusted, if the aspect is fulfilled:

| Aspect | Part of Standard | standardized description |
| ---- | ---- | ------ |
| Encryption | **Recommended** | **"[scs:encrypted]"** |
| Replication | **Recommended** | **"[scs:replicated]"** |

It is possible to use multiple of those aspects within one volume type. SCS will only ever look, if there is a volume type that has an aspect, but there don't have to be different volume types.
Example: one volume type that uses LUKS-encryption with a ceph storage with inherent replication would fulfill all recommendations of this standard.

## DEFAULT volume type

> There is always a default volume type defined in an OpenStack deployment. This volume type is created in the setup of cinder and will always be present in any OpenStack deployments under the name `__default__`. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from volume types named `__default__`.

Reviewer: ```suggestion
There is always a default volume type defined in an OpenStack deployment. This volume type is created in the setup of cinder and will always be present in any OpenStack deployments under the name `__default__`. This standard does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from volume types named `__default__`.
```

Author: i changed it.

> The parameters of volume types described in this standard do not have to be applied to the chosen default volume type. And the SCS will not make any assumptions about parameters being present in the default volume type.

Reviewer: ```suggestion
The parameters of volume types described in this standard do not have to be applied to the chosen default volume type. And this standard will not make any assumptions about parameters being present in the default volume type.
```

Author: I changed it.

## REQUIRED volume types

> Currently the SCS will not require volume types with certain specification. This will change in the future.

Reviewer: ```suggestion
Currently, this standard will not require volume types with certain specification. This will change in the future.
```

Author: I changed it.

## RECOMMENDED volume types

> The SCS recommends to have one or more volume types, that satisfy the need for encrpytion and replication.

Reviewer: ```suggestion
This standard recommends to have one or more volume types, that satisfy the need for encrpytion and replication.
```

Author: I changed it.

To fulfill this recommentation for now, the admin needs to add `[scs:replicated]` at the beginning or after any other scs aspect into the descriotion of the volume type.

### Example

One volume type that is configured as an encrypted volume type in a ceph backend, with automated replication would fit both recommendations and will be enough to comply to this part of the volume type standard.

It should look like the following part:

```text
+--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Field              | Value                                                                                                                                                        |
+--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
| access_project_ids | None                                                                                                                                                         |

> | description        | [scs:encrypted][scs:replicated] Content will be replicated three times to ensure consistency and availability for your data. LUKS encryption is used.        |

Author: i changed it.

features can not be discovered with full certainty by non-privileged users via the OpenStack API.

### Glossary

The following special terms are used throughout this standard document:
| Term | Meaning |
|---|---|
| volume | OpenStack ressource, virtual drive which usually resides in a network storage backend |
| AZ | Availability Zone |
| Volume QoS | Quality of Service object for Volumes |

## Motivation

As an SCS user, I want to be able to create volumes with certain common features, such as encryption or
replication, and to do so in a standardized manner as well as programmatically.
This standard outlines a way of formally advertising these common features for a volume type to
non-privileged users, so that the most suitable volume type can be discovered and selected easily — both by
the human user and by a program.

## Design Considerations

All Considerations can be looked up in detail in the [Decision Record for the Volume Type Standard.](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)

### Systematic Description of Volume Types

> To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. As for the time this standard is created, there is no way for users to discover all aspects through OpenStack commands. Therefore the aspects, that are fulfilled within a volume type, should be stated in the beginning of the **description** of a volume type in the following manner:

Reviewer 2: ack on `features` vs `aspects` - I like features more. but I would consider this a nit, as imho it's clear from the context that the same things are meant by these words.
It's also listed as synonymous in both merriam webster and oxford dictionary:

https://www.merriam-webster.com/thesaurus/aspect

> `[scs:aspect1, aspect2, ..., aspectN]...`

Reviewer 3: @anjastrunk Is `scs` a bit short (and meaningless)? As far as I recall, the participants in the original discussion didn't like `scs`.

> There is no sorting of aspects required. Every aspect should only be mentioned to the maximal amount of one.

Reviewer 3: Now we are back to no sorting? Why is that? As far as I recall, Anja suggested to use some sorting. I prefer sorting because it gives a canonical form which in the long run is easier to parse even for humans, because you don't have to read the whole thing to see if it has encryption (because, if it starts with replication, then it can't).

Reviewer 3: Besides, you copied my script verbatim, and the script still checks whether the list is sorted.

### Standardized Aspects

> The following table shows, which aspects are considered in this standard. The last column shows how the description of the volume type has to be adjusted, if the aspect is fulfilled:

Reviewer 3: I see the value of the additional column. The introductory paragraph is not very good though: the comma is wrong, and it's quite verbose. Why not use my proposal, which simply said

The following features are recognized.

Or if you must

The following features are recognized by this standard.

(But to my taste this is already a bit redundant; it's the point of the standard to state what's in the standard, so we need not repeat "in this standard" or "by this standard" all the time.)

| Aspect | Part of Standard | standardized description |
| ---- | ---- | ------ |
| Encryption | **Recommended** | **"[scs:encrypted]"** |

> | Replication | **Recommended** | **"[scs:replicated]"** |

Reviewer 3: I would still like to see the meaning column here.
The "Part of Standard" heading sounds quirky and needs to be renamed (or at the very least, make "standard" lower case)

> It is possible to use multiple of those aspects within one volume type. SCS will only ever look, if there is a volume type that has an aspect, but there don't have to be different volume types.
> Example: one volume type that uses LUKS-encryption with a ceph storage with inherent replication would fulfill all recommendations of this standard.

Reviewer 3: ```suggestion
It is possible that a single volume type combines multiple features.
For instance, one volume type that uses LUKS-encryption with a ceph storage with inherent replication would fulfill all recommendations of this standard.
```

## DEFAULT volume type

There is always a default volume type defined in an OpenStack deployment. This volume type is created in the setup of cinder and will always be present in any OpenStack deployments under the name `__default__`. This standard does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from volume types named `__default__`.

> The parameters of volume types described in this standard do not have to be applied to the chosen default volume type. And this standard will not make any assumptions about parameters being present in the default volume type.

Reviewer 3: I don't see the point in this: this assertion is entirely vapid. It's fine to say that no change of the default volume type is required by this standard, but this is only a side remark (because it follows from the text of the standard), and it warrants at most one sentence.

## REQUIRED volume types

> Currently, this standard will not require volume types with certain specification. This will change in the future.

Reviewer 3: I don't see the point in this. We always say that today's recommendations are tomorrow's requirements, and therefore all tests that we build MUST emit a warning, instead of a mere information, when some recommendation is violated. So this is vapid and can be dropped.

## RECOMMENDED volume types

> This standard recommends to have one or more volume types, that satisfy the need for encrpytion and replication.

Reviewer 3: The comma is wrong. I prefer the more clear and easily testable assertions from my proposal:

At least one volume type SHOULD feature encryption.

At least one volume type SHOULD feature replicaton.

Reviewer 3: One additional remark: 'satisfy the need for' -- satisfying needs is beyond this standard! We are talking about volume types having certain aspects or features. Whatever needs are satisfied by this (or not) is up the reader.

### Encryption

> There SHOULD at least be one volume type with the encryption aspect.

Reviewer 3: This sentence is redundant. The rest of the section belongs under a new section "Implemenation notes" (which should be a Supplement).

| description        | [scs:encrypted] This volume uses LUKS-encryption                                                                                                             |
| id                 | d63307fb-167a-4aa0-9066-66595ea9fb21                                                                                                                         |
| is_public          | True                                                                                                                                                         |
| name               | LUKS                                                                                                                                                         |
| qos_specs_id       | None                                                                                                                                                         |
+--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### Replication

> There SHOULD at least be one volume type with the replication aspect.

Reviewer 3: This sentence is redundant. The rest of the section belongs under a new section "Implemenation notes" (which should be a Supplement).

Replication states, whether or not there are multiple replicas of a volume. Thus answers the question, whether the data could survive a node outage. Unfortunately there are two ways replication can be achieved:

1. In the configuration of a volume type. It then is visible as extra_spec in the properties of a volume type.
2. Via the used backend. Ceph for example provides automatic replication, that does not need to be specified in the volume type. This is currently not visible for users.

> To fulfill this recommentation for now, the admin needs to add `[scs:replicated]` at the beginning or after any other scs aspect into the descriotion of the volume type.

Reviewer 3: This whole sentence is again redundant. How many places of this document shall the reader check for inconsistencies before they can proceed with the implementation???

> ### Example

Reviewer 3: I like this, but again, this belongs under a section "Implementation notes".

| id                 | d63307fb-167a-4aa0-9066-66595ea9fb21                                                                                                                         |
| is_public          | True                                                                                                                                                         |
| name               | hdd-three-replicas-LUKS                                                                                                                                      |
| properties         |                                                                                                                                                              |
| qos_specs_id       | None                                                                                                                                                         |
+--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

## OPTIONAL volume types

> Any other aspects of volume types, that can be found in the decision record are OPTIONAL. They SHOULD NOT be referenced in the way this standard describes. Some of them already are natively discoverable by users, while others could be described in the name or description of a volume type. Users should look into the provided volume types of the CSPs, if they want to use some of these other aspects.

Reviewer 3: Drop this paragraph. If it's unclear by now that other features are not recognized by this standard, and that arbitrary volume types may still exist, then we must improve the relevant places.

## Related Documents

> [Here is the decision record document.](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)

Reviewer 3: ```suggestion
- corresponding [decision record document](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)
```

don't say "here is" -- what does here even mean?

## Conformance Tests

> As there are currently no REQUIRED volume types, we can only look for the RECOMMENDED aspects and thus executing a conformance test is not a must.
> Furthermore the recommended aspects currently have to be described in the description by the deployer.
> Deployers MUST provided correct information, when adhering to this standard.
> The tests will only check the syntactical correctness.
> In future versions we aim to integrate some extra_specs for them in upstream OpenStack.
> It is also possible that a single volume type can currently fulfill all RECOMMENDED aspects.
> 
> The current test will check for the presence of `[encrypted]` and `[replicated]` in the description of at least one volume type.

Reviewer 3: ```suggestion
The script `/Tests/iaas/volume-types/volume-types-check.py` connects to an OpenStack environment and tests
the following:

- for each volume type: if its description starts with `[scs:....]`, then this prefix is a feature list
  (sorted, each entry at most once), and each entry is one of the possible features described here,
- the recommended volume types are present (otherwise, a WARNING is produced).

The return code is zero precisely when the test could be performed and the conditions are satisfied.
Otherwise, detailed errors and warnings are output to stderr.
```

---
title: Volume Type Standard
type: Standard
status: Draft
track:  IaaS
---

## Introduction

A volume is a virtual drive that is to be used by an instance (i. e., a virtual machine). With OpenStack,
each volume is created per a type that determines basic features of the volume as provided by the backend,
such as encryption, replication, or quality of service. As of the writing of this document, presence or absence of these
features can not be discovered with full certainty by non-privileged users via the OpenStack API.

### Glossary

The following special terms are used throughout this standard document:
| Term | Meaning |
|---|---|

> | volume | OpenStack ressource, virtual drive which usually resides in a network storage backend |

Reviewer 3: ```suggestion
| volume | OpenStack resource, virtual drive which usually resides in a network storage backend |
```

| volume feature | A certain feature a volume can possess |

> | aspect | part of a volume type that will activate a corresponding feature in a created volume |

Reviewer: ```suggestion
| aspect | Part of a volume type that will activate a corresponding feature in a created volume |
```

This standard outlines a way of formally advertising these common aspects for a volume type to
non-privileged users, so that the most suitable volume type can be discovered and selected easily — both by
the human user and by a program.

## Design Considerations

All considerations can be looked up in detail in the [Decision Record for the Volume Type Standard.](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)

### Systematic Description of Volume Types

To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. As for the time this standard is created, there is no way for users to discover all aspects through OpenStack commands. Therefore the aspects, that are fulfilled within a volume type, should be stated in the beginning of the **description** of a volume type in the following manner:

`[scs:aspect1, aspect2, ..., aspectN]...`

The mentioned aspects MUST be sorted alphebetically and every aspect should only be mentioned to the maximal amount of one.

### Standardized Aspects

> The following table shows, which aspects are considered in this standard. The last column shows how the description of the volume type has to be adjusted, if the aspect is fulfilled:

Reviewer 3: ```suggestion
The following table shows which aspects are considered in this standard. The third column shows how the description of the volume type has to be adjusted, if the aspect is fulfilled:
```

| Aspect | Requirement | standardized description | comment |
| ---- | ---- | ------ | ------ |

> | Encryption | **Recommended** | **"[scs:encrypted]"** | used to encrypt the volume |

Reviewer 3: ```suggestion
| Encryption | **Recommended** | **"[scs:encrypted]"** | volume is encrypted |
```
to make it consistent with the other comment, and to avoid confusion about the meaning of "used to"  (what is used? by whom?)

> | Replication | **Recommended** | **"[scs:replicated]"** | volume is replicated to aviod data loss in a case of hardware failure |

Reviewer 3: ```suggestion
| Replication | **Recommended** | **"[scs:replicated]"** | volume is replicated to avoid data loss in a case of hardware failure |
```

For instance, one volume type that uses LUKS-encryption with a ceph storage with inherent replication would fulfill all recommendations of this standard.

## DEFAULT volume type

There is always a default volume type defined in an OpenStack deployment. This volume type is created in the setup of cinder and will always be present in any OpenStack deployments under the name `__default__`. This standard does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from volume types named `__default__`.

The parameters of volume types described in this standard do not have to be applied to the chosen default volume type. And this standard will not make any assumptions about parameters being present in the default volume type.

## REQUIRED volume types

Currently, this standard will not require volume types with certain specification.

## RECOMMENDED volume types

This standard recommends to have one or more volume types, that feature encryption and replication.

## OPTIONAL volume types

> Any other aspects of volume types, that can be found in the decision record are OPTIONAL. They SHOULD NOT be referenced in the way this standard describes. Some of them already are natively discoverable by users, while others could be described in the name or description of a volume type. Users should look into the provided volume types of the CSPs, if they want to use some of these other aspects.

Reviewer: ```suggestion
Any other aspects of volume types, that can be found in the decision record are OPTIONAL. They SHOULD NOT be referenced in the way this standard describes. Some of them already are natively discoverable by users, while others could be described in the name or description of a volume type. Users should look into the provided volume types of the Cloud Service Providers, if they want to use some of these other aspects.
```

## Implementation Details

### Encryption

> Encryption for volumes is an option which has to be configured within the volume type. As an admin it is possible to set encryption-provider, key size, cipher and control location. Additionally to be discoverable by users an admin has to edit the description and add `[scs:encrypted]` at the beginning or after another scs aspect. It should look like this example:

Reviewer 3: This is not quite true anymore:

> add `[scs:encrypted]` at the beginning or after another scs aspect.

```suggestion
Encryption for volumes is an option which has to be configured within the volume type. As an admin it is possible to set encryption-provider, key size, cipher and control location. Additionally to be discoverable by users, the description should start with an aspect list such as `[scs:encrypted]` (potentially with additional aspects). It should look like this example:
```

| description        | [scs:encrypted] This volume uses LUKS-encryption                                                                                                             |
| id                 | d63307fb-167a-4aa0-9066-66595ea9fb21                                                                                                                         |
| is_public          | True                                                                                                                                                         |
| name               | LUKS                                                                                                                                                         |
| qos_specs_id       | None                                                                                                                                                         |
+--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### Replication

> Replication states, whether or not there are multiple replicas of a volume. Thus answers the question, whether the data could survive a node outage. Unfortunately there are two ways replication can be achieved:

Reviewer 3: ```suggestion
Replication states whether or not there are multiple replicas of a volume. Thus, it answers the question whether the data could survive a node outage. Unfortunately there are two ways replication can be achieved:
```

1. In the configuration of a volume type. It then is visible as extra_spec in the properties of a volume type.
2. Via the used backend. Ceph for example provides automatic replication, that does not need to be specified in the volume type. This is currently not visible for users.

> To fulfill this recommentation for now, the admin needs to add `[scs:replicated]` at the beginning or after any other scs aspect into the descriotion of the volume type.

Reviewer 3: ```suggestion
To fulfill this recommentation, the description should start with an aspect list such as  `[scs:replicated]` (potentially with additional aspects).
```
